File: week12/29/battleship.py
"""
<Battleships in a Board>

Given an m x n matrix board where each cell is a battleship 'X' or empty '.', return the number of the battleships on board.

Battleships can only be placed horizontally or vertically on board. In other words, they can only be made of the shape 1 x k (1 row, k columns) or k x 1 (k rows, 1 column), where k can be of any size. At least one horizontal or vertical cell separates between two battleships (i.e., there are no adjacent battleships).
"""

def countBattleships(board):
    if not board:
        return 0

    m, n = len(board), len(board[0])
    count = 0

    for i in range(m):
        for j in range(n):
            if board[i][j] == 'X':
                if i > 0 and board[i-1][j] == 'X':  # Check above
                    continue
                if j > 0 and board[i][j-1] == 'X':  # Check left
                    continue
                count += 1

    return count

board = [
    ['X', '.', '.', 'X'],
    ['.', '.', '.', 'X'],
    ['.', '.', '.', 'X']
]
print(countBattleships(board))  # 2

File: week12/25/battleship.cpp
#include <vector>
using namespace std;

class Solution {
public:
    int countBattleships(vector<vector<char>>& board) {
        int count = 0;
        int rows = board.size();
        int cols = board[0].size();

        for (int i = 0; i < rows; ++i) {
            for (int j = 0; j < cols; ++j) {
                // Check if this cell is a battleship
                if (board[i][j] == 'X') {
                    // Check if it's the start of a new battleship
                    if ((i == 0 || board[i - 1][j] != 'X') && 
                        (j == 0 || board[i][j - 1] != 'X')) {
                        count++;
                    }
                }
            }
        }

        return count;
    }
};